src/service/position-observer-impl.js

```diff
@@ -92,13 +124,32 @@ class AbstractPositionObserver {
           Math.floor(Math.random() * LOW_FIDELITY_FRAME_COUNT) : 0,
       trigger(position) {
         const prePos = entry.position;
+
         if (prePos
             && layoutRectEquals(prePos.positionRect, position.positionRect)
             && layoutRectEquals(prePos.viewportRect, position.viewportRect)) {
           // position doesn't change, do nothing.
           return;
         }
+
+        // Add the relative position regardless of whether the element
+        // is inside the view or not
+        position.relativePos = layoutRectsRelativePos(
+            position.positionRect, position.viewportRect
+        );
+
+        // Determine if the element is fully, partially or not contained
+        // within the viewport
+        position.visibility = layoutRectsIntersectionState(
+            position.positionRect, position.viewportRect
+        );
+
         if (layoutRectsOverlap(position.positionRect, position.viewportRect)) {
+          // Add the visible height of the element
+          position.visibleHeight = rectIntersection(
```

Comment:
I don't know if we should calculate `visibleHeight` here. We want positionObserver to provide element's position to the top-level viewport, but I don't think it's a good idea to ask positionObserver to do all the calculation. Registered element with positionRect and viewportRect can easily calculate its visible height themselves.